app/services/totp_creator.go

```diff
@@ -0,0 +1,48 @@
+package services
+
+import (
+	"bytes"
+	"image/png"
+
+	"github.com/keratin/authn-server/app/data"
+	"github.com/keratin/authn-server/lib/route"
+	"github.com/pkg/errors"
+	"github.com/pquerna/otp/totp"
+)
+
+// func TOTPCreator(
+// 	accountStore data.AccountStore, refreshTokenStore data.RefreshTokenStore, keyStore data.KeyStore, actives data.Actives, cfg *app.Config, reporter ops.ErrorReporter,
+// 	accountID int, audience *route.Domain, existingToken *models.RefreshToken,
+// ) (string, string, error) {
+
+//TOTPCreator handles the creation and storage of new TOTP tokens
+func TOTPCreator(accountStore data.AccountStore, totpCache data.TOTPCache, accountID int, audience *route.Domain) (*bytes.Buffer, error) {
+
+	account, err := AccountGetter(accountStore, accountID)
+	if err != nil {
+		return nil, err
+	}
+
+	//Generate totp key
+	key, err := totp.Generate(totp.GenerateOpts{
+		Issuer:      audience.Hostname,
+		AccountName: account.Username,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := totpCache.CacheTOTPSecret(account.ID, []byte(key.Secret())); err != nil {
+		return nil, errors.Wrap(err, "TOTPCreator")
+	}
+
+	//Convert to png
+	var buf bytes.Buffer
+	img, err := key.Image(200, 200)
+	if err != nil {
+		return nil, err
+	}
+	png.Encode(&buf, img)
+
+	return &buf, nil
```

Comment:
I'm not quite sure what you mean the raw url, could you elaborate a little more? Is that the login page url? 

Right now the URL for the AuthN instance and the username of the person is being encoded in the QR code and will show up in the authenticator app automatically when scanned. (Tested with Google Authenticator and Authy).

However I do think that this has to be changed anyways since there is currently no way to get the shared secret for entry if the QR code scanning is failing which is a mechanism allowed by authenticator apps. So on the initial `GET /totp` we could return the secret, the url and the username. With that information the client could render the QR code themselves, however I'd like an endpoint for it to render on the server (Maybe `GET /totp/qr`) since I sometimes have different frontends using different languages and I'd rather not have to get QR rendering working on every frontend.

Reply:
> However I do think that this has to be changed anyways since there is currently no way to get the shared secret for entry if the QR code scanning is failing which is a mechanism allowed by authenticator apps.

ah, yep, that's what i'm referring to. the value being encoded with QR [happens to be a URI](https://github.com/pquerna/otp/blob/master/totp/totp.go#L198) containing information like the secret.